app/src/main/java/com/github/se/travelpouch/model/documents/DocumentsManager.kt

```diff
@@ -0,0 +1,175 @@
+import kotlinx.coroutines.async
+import kotlinx.coroutines.flow.Flow
+import kotlinx.coroutines.flow.first
+import kotlinx.coroutines.flow.map
+import kotlinx.coroutines.tasks.await
+
+/** Helper class to download files from Firbase storage to local files */
+open class DocumentsManager(
+    private val contentResolver: ContentResolver,
+    private val storage: FirebaseStorage,
+    private val functions: FirebaseFunctions,
+    private val dataStore: DataStore<Preferences>,
+    private val thumbsDirectory: File
+) {
+  private val tag = DocumentsManager::class.java.simpleName
+
+  /**
+   * Return a Uri pointing to the file in the local storage.The file is firstly downloaded from
+   * Firebase storage if not already present.
+   *
+   * @param sourceMimeType The mimeType of the source file
+   * @param sourceTitle The title of the source file
+   * @param sourceRef The reference of the source file
+   * @param destinationFolder A pointer to the folder in which to store the file
+   */
+  fun getDocument(
+      sourceMimeType: String,
+      sourceTitle: String,
+      sourceRef: String,
+      destinationFolder: DocumentFile
+  ): Deferred<Uri> {
+    return CoroutineScope(Dispatchers.IO).async {
+      val existingFile = documentInCacheOrNull(sourceRef)
+      if (existingFile != null) {
+        Log.d(tag, "File already downloaded as $existingFile")
+        return@async existingFile
+      }
+
+      val file = destinationFolder.createFile(sourceMimeType, sourceTitle)
+
+      if (file == null) {
+        Log.e("DocumentViewModel", "Failed to create document file in specified directory")
+        throw Exception("Failed to create document file in specified directory")
+      }
+
+      Log.d(tag, "Downloading file to ${file.uri}")
+
+      val storageRef = storage.reference
+      val documentRef = storageRef.child(sourceRef)
+      val taskSnapshot = documentRef.stream.await()
+      contentResolver.openOutputStream(file.uri)?.use { outputStream ->
+        taskSnapshot.stream.use { it.copyTo(outputStream) }
+      }
+          ?: run {
+            Log.e(tag, "Failed to open output stream for URI: ${file.uri}")
+            throw Exception("Failed to open output stream for URI: ${file.uri}")
+          }
+      addDocumentRefToCache(sourceRef, file.uri.toString())
+      return@async file.uri
+    }
+  }
+
+  /**
+   * Return a Uri pointing to the file in the local storage if it is already present.
+   *
+   * @param sourceRef The reference of the source file
+   * @return The Uri pointing to the file in the local storage or null if the file is not present.
+   */
+  private suspend fun documentInCacheOrNull(sourceRef: String): Uri? {
+    val documentUid = stringPreferencesKey(sourceRef)
+    val pathFlow: Flow<String?> = dataStore.data.map { preferences -> preferences[documentUid] }
+    return try {
+      pathFlow.first()?.let {
+        Uri.parse(it).takeIf {
+          val afd = contentResolver.openAssetFileDescriptor(it, "r")
```

Comment:
There seems to be an issue with the way you handled the exceptions for some specific edge cases.

You can reliably reproduce the error by following these steps:
- Add a document and navigate to the preview in order to save the document in the cache
- Use the Files app to navigate to the folder and delete the file
- Go back to TravelPouch and try to display the preview, an exception is thrown and the preview will stay blank indefinitely

You will get the following stack trace:
```
2024-12-13 08:20:27.770  9669-9692  DocumentViewModel       com.github.se.travelpouch            E  Failed to download document
java.lang.IllegalArgumentException: Failed to determine if primary:TravelPouch Data/Scan 1734074384.jpg is child of primary:TravelPouch Data: java.io.FileNotFoundException: Missing file for primary:TravelPouch Data/Scan 1734074384.jpg at /storage/emulated/0/TravelPouch Data/Scan 1734074384.jpg
[...]
at com.github.se.travelpouch.model.documents.DocumentsManager.documentInCacheOrNull(DocumentsManager.kt:93)
```

Reply:
This is fixed. Nice catch